Update voice-chat.md (#1132)
Roblox/creator-docs: content/en-us/chat/voice-chat.md

---
title: Voice chat
description: Explains how to use the voice chat feature.
---

<Alert severity="success">
Voice chat is currently available to all 13+ [phone number verified](../production/publishing/account-verification.md#verify-through-phone-number) users in these countries: US, CA, GB, IE, AU, NZ, ES, MX, CL, CR, PR, FR, IT, AT, CHE, DE, JP, KR, CH, AR, COL, PE, DO, GT, UY, SV, HN, PY, NI, EC, BO, VE, PA, PT, and BR. Users **not** in these countries should use [ID verification](https://en.help.roblox.com/hc/en-us/articles/4407282410644) to enable chat with voice. Once verified, eligible 13+ phone verified users can opt‑in to use this feature by visiting their account **Settings** page or from within a voice enabled experience, allowing them to chat with voice in any Roblox experience that supports it. Experiences with voice often see an uplift in engagement, DAU, and spending.
</Alert>

> **Voice chat** is a feature that simulates realistic communication between yourself and other player characters. It is only available for places that support a maximum of 50 players.

Author: ```suggestion
**Voice chat** is a feature that enables real-time spoken communication between yourself and other player characters. It is only available for places that support a maximum of 50 players.
```

---
title: Voice Chat
description: Explains how to use the Voice Chat feature.
---

<Alert severity="success">

> Voice Chat is currently available to all 13+ [phone number verified](../production/publishing/account-verification.md#verify-through-phone-number) users in these countries: US, CA, GB, IE, AU, NZ, ES, MX, CL, CR, PR, FR, IT, AT, CHE, DE, JP, KR, CH, AR, COL, PE, DO, GT, UY, SV, HN, PY, NI, EC, BO, VE, PA, PT, and BR. Users **not** in these countries should use [ID verification](https://en.help.roblox.com/hc/en-us/articles/4407282410644) to enable chat with voice. Once verified, eligible 13+ phone verified users can opt‑in to use this feature by visiting their account **Settings** page or from within a voice enabled experience, allowing them to chat with voice in any Roblox experience that supports it. Experiences with voice often see an uplift in engagement, DAU, and spending.

Reviewer: ```suggestion
Voice Chat is currently available to all 13+ phone number verified users in a specific set of countries. Review the [Voice Chat FAQ](https://en.help.roblox.com/hc/en-us/articles/4405807645972-Voice-Chat-FAQs) to learn more. Users **not** in these countries should use [ID verification](https://en.help.roblox.com/hc/en-us/articles/4407282410644) to enable chat with voice. Once verified, eligible 13+ phone verified users can opt‑in to use this feature by visiting their account **Settings** page or from within a voice enabled experience, allowing them to chat with voice in any Roblox experience that supports it. Experiences with voice often see an uplift in engagement, DAU, and spending.
```

Reviewer: Adding this detail from other PR.

---
title: Voice Chat
description: Explains how to use the Voice Chat feature.
---

<Alert severity="success">

> Voice Chat is currently available to all 13+ phone number verified users in a specific set of countries. Review the [Voice Chat FAQ](https://en.help.roblox.com/hc/en-us/articles/4405807645972-Voice-Chat-FAQs) to learn more. Users **not** in these countries should use [ID verification](https://en.help.roblox.com/hc/en-us/articles/4407282410644) to enable chat with voice. Once verified, eligible 13+ phone verified users can opt‑in to use this feature by visiting their account **Settings** page or from within a voice enabled experience, allowing them to chat with voice in any Roblox experience that supports it. Experiences with voice often see an uplift in engagement, DAU, and spending.

Reviewer 2: tysm

<img src="../assets/players/voice-chat/In-Experience-Example.jpg" width="800" alt="Two users chatting with voice inside an experience" />

## Enable Voice Chat

Before you can enable Voice Chat in an experience, you must first [publish](../production/publishing/publish-experiences-and-places.md) it, then:

1. Open your experience in Studio.
1. Open **File**&nbsp;⟩ **Game Settings**.
1. Navigate to the **Communication** tab on the left side of the window.
1. Toggle **Enable Microphone** so the selector turns from gray to green.
1. <Chip label="OPTIONAL" size="small" variant="outlined" /> For greater communication among players within your experience, toggle on **Enable&nbsp;Camera** to allow eligible players to animate their avatar with their movement.
1. [Publish](../production/publishing/publish-experiences-and-places.md) the place to save the changes.

Voice Chat will now be available to verified 13+ users who opt‑in to the feature, in every place within the experience that's set to a maximum of 50 players.

### Set maximum players

> If you previously set the maximum number of players in a place to more than 50, you'll need to reduce it to support .

Reviewer: ```suggestion
If you previously set the maximum number of players in a place to more than 50, you'll need to reduce it to support Voice Chat.
```

Reviewer: I think this got accidentally cut...

	-- This is provided for you if VoiceChatService.EnableDefaultVoice is true
	-- May need to be reworked for generality if your place puts AudioDeviceInput objects elsewhere
	return forPlayer:FindFirstChildWhichIsA("AudioDeviceInput")
end

local function onTeamChanged(player : Player)
	local team = player.Team
	if not team then return end

	local device = findAudioInput(player)
	if not device then return end
	
	-- Only permit teammates to hear each other
	device.AccessType = Enum.AccessModifierType.Allow
	local allowed = {}
	for _, player : Player in team:GetPlayers() do
		table.insert(allowed, player.UserId)
	end

	device:SetUserIdAccessList(allowed)
end

local function onTeamAdded(team : Team)
	team.PlayerAdded:Connect(onTeamChanged)
	team.PlayerRemoved:Connect(onTeamChanged)
end

for _, team : Team in Teams:GetTeams() do
	onTeamAdded(team)
end

Teams.ChildAdded:Connect(function(child : Instance)
	if child:IsA("Team") then
		onTeamAdded(child)
	end
end)
```

#### Non-spatial

You might not want to use the default, proximity-chat behavior. Depending on your experience's needs, it might make more sense to hear others with equal volume, regardless of their geographic location.

> If you [disable default Voice Chat](#disable-per-place), you can then implement flat/nonspatial Voice Chat through the following `Class.Script` within `Class.ServerScriptService`:

Reviewer: Just making it a bit more concise...

```lua title="Non-Proximity Chat"
local Players = game:GetService("Players")
local Workspace = game:GetService("Workspace")

local function wireUp(source : Instance, target : Instance)
	local wire = Instance.new("Wire")
	wire.SourceInstance = source
	wire.TargetInstance = target
	wire.Parent = source
end

-- Set up a global volume slider for everybody's voice
local volumeSlider = Instance.new("AudioFader", Workspace)
local output = Instance.new("AudioDeviceOutput", Workspace)
wireUp(volumeSlider, output)

local function onPlayerAdded(player : Player)
	local device = Instance.new("AudioDeviceInput", player)
	device.Player = player
	-- Route each new player's microphone input to the global volume slider
	wireUp(device, volumeSlider)
end

for _, player in Players:GetPlayers() do
	onPlayerAdded(player)
end
Players.PlayerAdded:Connect(onPlayerAdded)
```

> The above script runs on the server; it creates an `Class.AudioDeviceInput|AudioDeviceInput` per-player, and connects them all to a shared `Class.AudioFader|AudioFader`.
> 
> Taking this behavior as-is allows each player to hear everybody – including themselves! Hearing yourself can be quite distracting: you can place this `Class.LocalScript` in `Class.StarterPlayerScripts` to remove the `Class.Wire|Wires` that would otherwise ferry your own voice back to your own speakers:

Reviewer: ```suggestion
<Alert severity="warning">
The above script runs on the server, creating an `Class.AudioDeviceInput` per‑player and connecting them all to a shared `Class.AudioFader`. Taking this behavior as‑is allows players to hear all other players **and** themselves. Since hearing yourself can be quite distracting, you can place the following `Class.LocalScript` in `Class.StarterPlayerScripts` to remove the `Class.Wire|Wires` that would otherwise ferry your own voice back to your own speakers.
</Alert>
```

Reviewer: I think it's worth putting this inside a yellow alert box for increased visibility.

Voice chat is currently available to all 13+ [phone number verified](../production/publishing/account-verification.md#verify-through-phone-number) users in these countries: US, CA, GB, IE, AU, NZ, ES, MX, CL, CR, PR, FR, IT, AT, CHE, DE, JP, KR, CH, AR, COL, PE, DO, GT, UY, SV, HN, PY, NI, EC, BO, VE, PA, PT, and BR. Users **not** in these countries should use [ID verification](https://en.help.roblox.com/hc/en-us/articles/4407282410644) to enable chat with voice. Once verified, eligible 13+ phone verified users can opt‑in to use this feature by visiting their account **Settings** page or from within a voice enabled experience, allowing them to chat with voice in any Roblox experience that supports it. Experiences with voice often see an uplift in engagement, DAU, and spending.
</Alert>

By default, **Voice chat** is proximity-based, adjusting the volume of participants based on how close you are to other users who are speaking. The closer you are to another user's avatar, the louder their voice; conversely, the farther away you are, the softer their voice.

> However, by using the **Audio API**, you can customize how voice chat is presented in your experience.

Reviewer 3: Should Audio API hyperlink somewhere?

Author: It could hyperlink to the devforum announcement, but there's so much _other_ (non-voice-specific) functionality in there that I'm not sure it would be more readable honestly :\

Reviewer: I'd prefer no links to DevForum posts. And AFAIK, "Audio API" is not really a proper "thing" but a set of things controlled by other APIs, so there's not a great singular place to link.

Reviewer 3: sounds like we might need an Audio API page

Author: We have one, but it's both not-specific-to-voice, and contains descriptions for Sound/SoundGroup/SoundEffects as well

1. [Publish](../production/publishing/publish-experiences-and-places.md) the place to save the changes.

Voice chat will now be available to verified 13+ users who opt‑in to the feature, in every place within the experience that's set to a maximum of 50 players.

### Set maximum players

If you previously set the maximum number of players in a place to more than 50, you'll need to reduce it to support voice chat.

1. In the left-hand navigation of the **Game Settings** window, select **Places**. Every place within your experience displays.
1. Click the **&ctdot;** button next to the place with more than 50 players, then select **Configure Place**.
1. In the **Max Players** field, enter any number less than or equal to 50.
1. Click the **Save** button and then [publish](../production/publishing/publish-experiences-and-places.md) to save the changes.

When you update the maximum number of players in a place to fewer than 50, there may be servers already configured to a different, higher number. Since those servers won't support voice chat, it's recommended to [restart servers](../production/publishing/publish-experiences-and-places.md#update-experiences).

> ### Customize voice transmission

Reviewer: ```suggestion
### Customize voice behavior
```
```suggestion
### Customize voice transmission
```

Reviewer: Behavior vs. transmission?

Author: Behavior looks good to me

1. Right‑click over any visible service in the **Explorer** window and select **Show&nbsp;Services…** from the context menu.
2. Select `Class.VoiceChatService` in the popup window and click **Insert**. The service appears in the **Explorer** hierarchy.

   <img src="../assets/studio/explorer/VoiceChatService.png" width="320" alt="VoiceChatService in Explorer hierarchy" />

3. With `Class.VoiceChatService` selected, set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` in the **Properties** window.

#### Team-based

To implement team-based chat where only teammates can hear one another, you can use the following `Class.Script` within `Class.ServerScriptService`:

```lua title="Script"
local Teams = game:GetService("Teams")

local function findAudioInput(forPlayer : Player) : AudioDeviceInput?
	-- Assumes there is only one AudioDeviceInput per player, parented to the Player object

> 	-- May need to be reworked for generality if your place does not match this

Author: ```suggestion
	-- This is provided for you if `VoiceChatService.EnableDefaultVoice` is true,
	-- but this may need to be reworked for generality if your place puts `AudioDeviceInput`s elsewhere
```

	return forPlayer:FindFirstChildWhichIsA("AudioDeviceInput")
end

local function onTeamChanged(player : Player)
	local team = player.Team
	if not team then return end

	local device = findAudioInput(player)
	if not device then return end
	
	-- Only permit teammates to hear each other
	device.AccessType = Enum.AccessModifierType.Allow
	local allowed = {}
	for _, player : Player in team:GetPlayers() do
		table.insert(allowed, player.UserId)
	end

	device:SetUserIdAccessList(allowed)
end

local function onTeamAdded(team : Team)
	team.PlayerAdded:Connect(onTeamChanged)
	team.PlayerRemoved:Connect(onTeamChanged)
end

for _, team : Team in Teams:GetTeams() do
	onTeamAdded(team)
end

Teams.ChildAdded:Connect(function(child : Instance)
	if child:IsA("Team") then
		onTeamAdded(child)
	end
end)
```

#### Non-spatial

If you [disable default voice chat](#disable-per-place), you can then implement flat/nonspatial voice chat through the following `Class.Script` within `Class.ServerScriptService`:

> ```lua title="Script"

Author: ```suggestion
```lua title="Non-proximity Chat"
```

1. Open **File**&nbsp;⟩ **Game Settings**.
1. Navigate to the **Communication** tab on the left side of the window.
1. Toggle **Enable Microphone** so the selector turns from gray to green.

> 1. <Chip label="OPTIONAL" size="small" variant="outlined" /> For greater communication among players within your experience, toggle on **Enable&nbsp;Camera** to allow eligible players to animate their avatar with their movement.

Author: I'd probably go even a step further and use backticks
```
`Class.Player|Players` 
```
to link to the class documentation

> 1. <Chip label="OPTIONAL" size="small" variant="outlined" /> For greater communication among players within your experience, toggle on **Enable&nbsp;Camera** to allow eligible players to animate their avatar with their movement.

Reviewer: For context over many years, at some point "users" was largely subbed in for "players" in certain contexts, but we still use the terms generally interchangeably in docs. The term "players" sounds a bit less formal and many APIs (including the actual Players service and Player objects) sync with that less-formal tone.

1. [Publish](../production/publishing/publish-experiences-and-places.md) the place to save the changes.

> Voice chat will now be available to verified 13+ users who opt‑in to the feature, in every place within the experience that's set to a maximum of 50 players.

Author: ```suggestion
Voice Chat will now be available to verified 13+ users who opt‑in to the feature, in every place within the experience that's set to a maximum of 50 players.
```

### Set maximum players

> If you previously set the maximum number of players in a place to more than 50, you'll need to reduce it to support voice chat.

Author: ```suggestion
If you previously set the maximum number of players in a place to more than 50, you'll need to reduce it to support Voice Chat.
```

1. In the left-hand navigation of the **Game Settings** window, select **Places**. Every place within your experience displays.
1. Click the **&ctdot;** button next to the place with more than 50 players, then select **Configure Place**.
1. In the **Max Players** field, enter any number less than or equal to 50.
1. Click the **Save** button and then [publish](../production/publishing/publish-experiences-and-places.md) to save the changes.

When you update the maximum number of players in a place to fewer than 50, there may be servers already configured to a different, higher number. Since those servers won't support voice chat, it's recommended to [restart servers](../production/publishing/publish-experiences-and-places.md#update-experiences).

### Customize voice behavior

> Voice chat is **proximity-based** by default, adjusting the volume of participants based on how close they are to each other. However, you can set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` to take control over how voices are set up and used in your experience:

Reviewer 2: the default rolloff curve 

people shouldn't be audible 80 studs away 

with inverse you can hear people but it's quiet

> Voice chat is **proximity-based** by default, adjusting the volume of participants based on how close they are to each other. However, you can set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` to take control over how voices are set up and used in your experience:

Reviewer 2: voice audio curve versus audio api one 

new: never fully gets to 0 AH

> Voice chat is **proximity-based** by default, adjusting the volume of participants based on how close they are to each other. However, you can set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` to take control over how voices are set up and used in your experience:

Reviewer 2: voice chat: attenuation: legacy old one 
inverse you can make quieter 

you can hear things far away with inverse, "universal"

> 1. Right‑click over any visible service in the **Explorer** window and select **Show&nbsp;Services…** from the context menu.

Author: I think we should also clarify that VoiceChatService may already be instantiated for you

2. Select `Class.VoiceChatService` in the popup window and click **Insert**. The service appears in the **Explorer** hierarchy.

   <img src="../assets/studio/explorer/VoiceChatService.png" width="320" alt="VoiceChatService in Explorer hierarchy" />

> 3. With `Class.VoiceChatService` selected, set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` in the **Properties** window.

Author: In templates yes, but not necessarily in existing placefiles

> 3. With `Class.VoiceChatService` selected, set `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enum.AudioApiRollout|Enabled` in the **Properties** window.

Reviewer 2: previous experience don't have it enabled, new ones do

#### Team-based

> To implement team-based chat where only teammates can hear one another, you can use the following `Class.Script` within `Class.ServerScriptService`:

Author: ```suggestion
To implement `Class.Team`-based chat – where only teammates can hear one another – you can use the following `Class.Script` within `Class.ServerScriptService`:
```

When you update the maximum number of users in a place to fewer than 50, there may be servers already configured to a different, higher number. Since those servers won't support voice chat, it's recommended to [restart servers](../production/publishing/publish-experiences-and-places.md#update-experiences).

### Using the Audio API to control Voice Chat

By setting `Class.VoiceChatService.UseAudioApi|UseAudioApi` to `Enabled`, participant voices will be controlled by `Class.AudioDeviceInput|AudioDeviceInputs`.

Scripts can use the properties and methods of `Class.AudioDeviceInput|AudioDeviceInputs` to implement custom, game-specific logic.

For example, if you wanted to implement team-based chat – where only teammates can hear one another – you could do something along the lines of

```lua
local function findAudioInput(forPlayer : Player) : AudioDeviceInput?
	-- Assumes that there is only one AudioDeviceInput per player
	-- and that it is parented to the Player object
	-- may need to be reworked for generality if your scene does not match this
	return forPlayer:FindFirstChildWhichIsA("AudioDeviceInput")

> end

Reviewer 3: I think most ppl will want a full working example, which might be lengthier than we'd like in this overall page.
For TextChat, we have a full examples subpage where we go into greater detail on the code samples.

For here, I think it would be beneficial to use this sample to onboard users to the AudioApi in general. IIRC, users would need to supply their own AudioDeviceInput for this to work.

local function onTeamChanged(player : Player)
	local team = player.Team
	if not team then
		return
	end

	local device = findAudioInput(player)
	if not device then
		return
	end
	
	-- Only permit teammates to hear each other
	device.AccessType = Enum.AccessModifierType.Allow
	local allowed = {}
	for _, player : Player in team:GetPlayers() do
		table.insert(allowed, player.UserId)
	end

	device:SetUserIdAccessList(allowed)
end

local function onTeamAdded(team : Team)
	team.PlayerAdded:Connect(onTeamChanged)
	team.PlayerRemoved:Connect(onTeamChanged)
end

local Teams : Teams = game:GetService("Teams")

for _, team : Team in Teams:GetTeams() do
	onTeamAdded(team)
end

Teams.ChildAdded:Connect(function(child : Instance)
	if child:IsA("Team") then
		onTeamAdded(child)
	end
end)
```

> By setting `Class.VoiceChatService.EnableDefaultVoice|EnableDefaultVoice` to **false**, no default voice chat setup will be provided – this can be used to implement "flat"/nonspatial voice chat (instead of the standard proximity chat)

Reviewer 3: Im confused reading this. What are the expected permutations between UseAudioApi true/false * EnableDefaultVoice true/false?

Author: The two properties are conveying two different things

`EnableDefaultVoice` means you get the "default" voice setup. Peoples' voices come from their heads.

`UseAudioApi` means voice chat is implemented *using* the audio api; whether you take the default setup or not. So when UseAudioApi is Enabled, peoples' microphones are represented/controlled by `AudioDeviceInput`s, instead of being hidden/internal.

So, you can have 4 possible combinations of these:
1. Default voice setup, but not using the audio api – voice chat happens, and you can't customize it any further
2. Default voice setup, but using the audio api – voice chat happens, using an AudioDeviceInput for each player's microphone and an AudioEmitter to beam their voice out into the world from their avatar
3. Non-default voice setup, not using the audio api – voice chat is effectively disabled
4. Non-default voice setup, using the audio api – no voice chat is provided for you, but you *could* roll your own with the audio API

Author: UseAudioApi will eventually go away – it's a 3-phase rollout

Reviewer 3: Do you think its worth trying to distill your comment onto the docs page if UseAudioApi will stick around for some time?

```lua
-- Ensure VoiceChatService.EnableDefaultVoice is **false**
-- since this script will be doing something custom instead

local function wireUp(source : Instance, target : Instance)
	local wire = Instance.new("Wire")
	wire.SourceInstance = source
	wire.TargetInstance = target
	wire.Parent = source
end

-- Set up a global volume slider for everybody's voice
-- so that it's easy for other code to adjust its volume
local volumeSlider = Instance.new("AudioFader", workspace)
local output = Instance.new("AudioDeviceOutput", workspace)
wireUp(volumeSlider, output)

local function onPlayerAdded(player : Player)
	local device = Instance.new("AudioDeviceInput", player)
	device.Player = player
	-- Route each new player's microphone input to the global volume slider
	wireUp(device, volumeSlider)
end

local Players = game:GetService("Players")
for _, player in Players:GetPlayers() do
	onPlayerAdded(player)
end
Players.PlayerAdded:Connect(onPlayerAdded)
```

You may not want players to hear themselves; in that case, the above script should be combined with a local script along the lines of

> ```

Author: ```suggestion
```lua
```
